### DISCUSSION TITLE
typescript loses state node type enum inference

### DISCUSSION THREAD

**Discussion by mong8se** [statelyai/xstate#2624] [Q&A]
Sorry for the word soup title, it's a really hard thing to search for, because of the overloading of the word "type" in the context of this problem. The typescript type of the state node type is the issue.

See below, the config is exactly the same, but typescript complains about `machine2`, it seemingly loses connection to the enum on the state node type:

>  Type 'string' is not assignable to type '"final" | "atomic" | "compound" | "parallel" | "history" | undefined'.

`machine3` removes the complaint, but it's redundant and it also removes a lot of the type checking on the machine, can misspell the `finish` event for example, and `machine3` won't complain.

Is there a better solution that still keeps the config as pure json? The goal is to `import config from './config.json'`

```typescript
import { MachineConfig, createMachine } from "xstate";

type AppContext = {};

type AppEvent = { type: "finish" };

export const machine1 = createMachine<AppContext, AppEvent>({
  context: {},
  initial: "idle",
  states: {
    idle: {
      on: {
        finish: "finished",
      },
    },
    finish: {
      entry: "finalAction",
      type: "final",
    },
  },
});

export const sameConfigExternalized = {
  context: {},
  initial: "idle",
  states: {
    idle: {
      on: {
        finish: "finished",
      },
    },
    finish: {
      entry: "finalAction",
      type: "final",
    },
  },
};

export const machine2 = createMachine<AppContext, AppEvent>(
  sameConfigExternalized
);

export const machine3 = createMachine<AppContext, AppEvent>(
  sameConfigExternalized as MachineConfig<AppContext, any, AppEvent>
);
```

**Comment by ChrisShank** [comment]
This should fix your TS errors:
```ts
export const sameConfigExternalized = {
  context: {},
  initial: "idle",
  states: {
    idle: {
      on: {
        finish: "finished",
      },
    },
    finish: {
      entry: "finalAction",
      type: "final",
    },
  },
} as const;
```

**Reply by mong8se** [reply]
Thanks! That does stop the "final" error, but it's not the final error. Now it complains that the context is read-only. For example if I have an array in context:

```
      Type '{ readonly values: readonly []; }' is not assignable to type 'AppContext'.
        Types of property 'values' are incompatible.
          The type 'readonly []' is 'readonly' and cannot be assigned to the mutable type '[]'.
```

Is there a way to use `const` and context too?

Appreciate your time!

```typescript
import { createMachine } from "xstate";

type AppContext = { values: [] };

type AppEvent = { type: "finish" };

export const machine1 = createMachine<AppContext, AppEvent>({
  context: { values: [] },
  initial: "idle",
  states: {
    idle: {
      on: {
        finish: "finished",
      },
    },
    finish: {
      entry: "finalAction",
      type: "final",
    },
  },
});

export const sameConfigExternalized = {
  context: { values: [] },
  initial: "idle",
  states: {
    idle: {
      on: {
        finish: "finished",
      },
    },
    finish: {
      entry: "finalAction",
      type: "final",
    },
  },
} as const;

export const machine2 = createMachine<AppContext, AppEvent>(
  sameConfigExternalized
);
```

**Reply by ChrisShank** [reply]
Oooh didnt think of that edge case. 🤔

the best TS experience is going to be inlining the machine in `createMachine`! If we take a step back is there a reason you are trying to define the config else where?

**Reply by mong8se** [reply]
The main reason is so I can also import the config into a script which uses `xstate-plantuml` package to auto generate an svg visualization of the machine.

The other way would be to have the script be able to import the `.ts` file directly and get `machine.config`, but I'm having a heck of a time having a simple script be able to boot the entire react stack to import the machine to get the config, so I was hoping for a more lightweight solution with the externalized json config.

Maybe I'm over thinking it, is there a simpler solution?

The secondary reason is to be able to split up big configs later, for example break out all the top level parallel states into separate files and compose them into one smaller machine index.ts

Appreciate your time, Chris.
